### ISSUE TITLE
Fix `postinstall` and finish `trustedDependencies`

### ISSUE THREAD

**Issue by Electroid** [oven-sh/bun#4959]
## 🔍 Summary

Bun introduces the concept of [`trustedDependencies`](https://bun.sh/docs/cli/install#trusted-dependencies), which allows you to specify an allow-list of packages that are allowed to run lifecycle scripts, such as `postinstall`. This is because dependencies can run arbitrary shell scripts when installed, and that's generally not great.

package.json:
```json
{
  "dependencies": {
    "utf-8-validate": "^6.0.3"
  },
  "trustedDependencies": [
    "utf-8-validate"
  ]
}
```

However, there are dependencies that you need to _trust_ to run scripts, because otherwise they do not work. To do this, you can add the package name to the `trustedDependencies` list in your package.json.

But doing this for every package would be annoying, so Bun would also have a default allow-list of the most popular packages on npm that need to run scripts, like utf-8-validate or esbuild. https://github.com/oven-sh/bun/pull/4263

We planned to include these changes in 1.0, but ran out of time. Once Bun supports these changes, you wouldn't need to use `trustedDependencies` for 99% of the time, except for more obscure dependencies. Also, when running `bun add` or `bun install`, there would be a prompt to indicate that the dependency is using scripts.

## 🚧 Workaround

1. Add package to `trustedDependencies` in your package.json.
```json
{
  "trustedDependencies": ["utf-8-validate"]
}
```

2. Run `bun remove`, then `bun add`, to ensure script is re-run.
```sh
bun remove utf-8-validate
bun add utf-8-validate
```

**Comment by birkskyum** [comment]
Also, it would be great to expand the test suite here beyond a clean install, because it appear to [behave incorrectly](https://github.com/oven-sh/bun/issues/4549). I've made a PR here that does test more scenarios:
- https://github.com/oven-sh/bun/pull/4629

Scenarios that aren't covered in that PR yet is if a lockfile i.e. already exist, and then the library is added to the trustedDependencies array, followed by `bun install`, or the other way around, but it's a good starting point.

**Comment by intrnl** [comment]
Are the defaults for trustedDependencies overrideable?

**Comment by ImBIOS** [comment]
I deleted bun.lockb and node_modules, then run `bun i`, all trustedDependencies works like charm.

**Comment by labibimamhidayat** [comment]
> I deleted bun.lockb and node_modules, then run `bun i`, all trustedDependencies works like charm.

this solve my problem, thank youu

**Comment by uncenter** [comment]
Any plans to prompt the user to approve/deny a package's postinstall script during the install command?

**Comment by taep96** [comment]
> Any plans to prompt the user to approve/deny a package's postinstall script during the install command?

And add it to trustedDependencies, error if running on CI?

**Comment by elmpp** [comment]
Please note that when the postinstall scripts are run, the process.arch values are being influenced by bun. 

There is a repro repo here - https://github.com/oven-sh/bun/issues/6367#issuecomment-1752002238

**Comment by brianmhunt** [comment]
The `trustedDependencies` is a great idea to help with the supply chain security vectors.

Having it turned on by default out of the gate is a hard blocker to using bun for large pre-existing codebases that may have hundreds of `package.json` files with many thousands of packages, often from private repos.

A few things would help here:

1. a `--trust-all-dependencies` flag that circumvents the `trustedDependencies` logic entirely
2. a glob on the `trustedDependencies` for groups e.g. `@firebase/*`
3. some way to ignore packages from inside the current workspace / monorepo (it may work already, just flagging for thought)

Ideally we'd all move to a safe set of `trustedDependencies` but until that's practical it'd be nice to work around this issue to be able to use bun for package management.

**Comment by uncenter** [comment]
Those are some great suggestions! Big fan of the first one.

**Comment by brianmhunt** [comment]
Thanks @uncenter.

One thing I should have noted is that `--trust-all-dependencies` is a proper subset of a glob-based CLI arg like `--trust-dependencies='*'`. The prior is probably simpler and faster to delivery, but having both expands the argument-surface. 

Given that it's a new mechanism in the ecosystem, people aren't going to know to look for `trustedDependencies`.  I didn't, I only got here by googling "postinstall bun.sh".  To save others pain/churn, it may be worthwhile to emit a message to the CLI when:

1. a `package.json` of a package being installed has postinstall/etc items that would ordinarily run but don't because they're not in `trustedDependencies`
2. the root `package.json` where `bun i` is being run does not have a `trustedDependencies`

Perhaps it's even worth printing a list of every package whose postinstall didn't run.  That's less of a problem/value when people know the mechanism exists, but the visibility is still important for package installation diagnostics.

**Comment by uncenter** [comment]
Your last suggestion there is a must have! Either that or something more interactive like what I suggested:

> Any plans to prompt the user to approve/deny a package's postinstall script during the install command?

**Comment by brianmhunt** [comment]
A patch like this might get it working for the simplest solution:

```patch
❯ git diff --patch
diff --git a/src/install/install.zig b/src/install/install.zig
index b4650cb75..75dbfb6ab 100644
--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -6387,6 +6387,7 @@ pub const PackageManager = struct {
         clap.parseParam("--cwd <STR>                       Set a specific cwd") catch unreachable,
         clap.parseParam("--backend <STR>                   Platform-specific optimizations for installing dependencies. " ++ platform_specific_backend_label) catch unreachable,
         clap.parseParam("--link-native-bins <STR>...       Link \"bin\" from a matching platform-specific \"optionalDependencies\" instead. Default: esbuild, turbo") catch unreachable,
+        clap.parseParam("--trust-all-dependencies          Run scripts for all dependencies (ignore trustedDependencies in package.json)") catch unreachable,
 
         // clap.parseParam("--omit <STR>...                   Skip installing dependencies of a certain type. \"dev\", \"optional\", or \"peer\"") catch unreachable,
         // clap.parseParam("--no-dedupe                       Disable automatic downgrading of dependencies that would otherwise cause unnecessary duplicate package versions ($BUN_CONFIG_NO_DEDUPLICATE)") catch unreachable,
@@ -6456,6 +6457,7 @@ pub const PackageManager = struct {
         no_verify: bool = false,
         ignore_scripts: bool = false,
         no_summary: bool = false,
+        trust_all_dependencies: bool = false,
 
         link_native_bins: []const string = &[_]string{},
 
@@ -6528,6 +6530,7 @@ pub const PackageManager = struct {
             cli.verbose = args.flag("--verbose");
             cli.ignore_scripts = args.flag("--ignore-scripts");
             cli.no_summary = args.flag("--no-summary");
+            cli.trust_all_dependencies = args.flag("--trust-all-dependencies")
 
             // link and unlink default to not saving, all others default to
             // saving.
@@ -7469,7 +7472,7 @@ pub const PackageManager = struct {
                             }
                         }
 
-                        if (resolution.tag == .workspace or this.lockfile.trusted_dependencies.contains(@as(u32, @truncate(String.Builder.stringHash(name))))) {
+                        if (resolution.tag == .workspace or cli.trust_all_dependencies or this.lockfile.trusted_dependencies.contains(@as(u32, @truncate(String.Builder.stringHash(name))))) {
                             var scripts = this.lockfile.packages.items(.scripts)[package_id];
                             if (scripts.hasAny()) {
                                 var path_buf: [bun.MAX_PATH_BYTES]u8 = undefined;
```

**Comment by brianmhunt** [comment]
I did a little canvassing of developers and I think there's a few things to note here:

1. many first impressions with bun are going to be `bun i` and because it's not going to work the same as `npm i` it will silently, inexplicably not work for many
2. the early negative experience will curtail early adoption and engagement
3. low engagement means the 👍 count on this is going to be greatly understated (notable because 👍 seems to be a barometer on the post 1.0 patch notes)
4. while #5077 ostensibly fixes this (as a side effect) it doesn't _actually_ fix it for the use-cases described here
5. #5077 is a complex PR, yet a trivial PR could resolve this
6. while `trustedDependencies` is a great idea, it deviates from bun's stated desire to be compatible with Node.js

This issue strikes me as a silent killer of adoption, so I'd be inclined to look at it sooner rather than later.  Tagging @Electroid @Jarred-Sumner for visibility.

**Comment by Jarred-Sumner** [comment]
@dylan-conway is actively focused on addressing this in https://github.com/oven-sh/bun/commits/postinstall_3. We will add a flag to allow all as well. We are aiming to ship it by Monday. 

Part of the delay is because we must run these postinstall scripts in parallel or it will make bun install slow. Not only that, but it has so much potential to impact bun install’s performance we probably will add a separate postinstall caching step in the future. We didn’t have a completely reliable 100% of the time way to invoke multiple subprocesses in parallel on macOS until a couple weeks ago (there were longstanding 1 out of 100,000 or so bugs that we addressed recently)

**Comment by brianmhunt** [comment]
@Jarred-Sumner given the (truly) epic scope of this project, I'd be surprised if it was only 100,000! 😉

**Comment by fdaciuk** [comment]
The flag idea to allow all is really great! @Jarred-Sumner can you make this option configurable via `$HOME/.bunfig.toml`?

**Comment by brianmhunt** [comment]
If the perf branch is taking a long time, is it worth breaking the flag into a separate story?  The analysis is very different from the performance issue, namely:

| Perspective | `--trust-all-dependencies` Flag | Parallelized install |
| -- | -- | -- |
| Complexity of fix | Trivial | Complex |
| Impact on affected packages | Silent failure to install | Slower installation |
| First user experience | Inexplicable deviation from expectation and unusable installation | Performance at-worst equivalent to npm/etc |
| Required user remediation | Time consuming investigation [1] | N/A |

[1] The investigation entails:
1. discovering why `bun i` failed (i.e. google + repo search, some knowledge of postinstall, etc.)
3. finding out what postinstall/etc didn't run
4. editing `package.json` files to add `trustedDependencies`

There's low hanging fruit in the flag (and perhaps logging) to fix a very negative experience, and I fear it being conflated with a performance impression may be lowering the apparent significance of this issue.

cc @Jarred-Sumner @Electroid @dylan-conway

**Comment by strokirk** [comment]
I would like to second the suggestion to print a list of ignored post-install scripts. Currently there isn't much guidance about this issue on the internet, so while trying to install `vue-demi` today it took quite a bit of digging to realize that `bun` doesn't run `postinstall` scripts, and even after adding the script to `trustedDependencies`, `bun i` didn't run the script.

**Comment by uncenter** [comment]
#7132 fixes the big issue here but leaves out some ideas that have been thrown around:

- Prompt users to accept/deny execution of each post-install script OR just print a list of ignored post-install scripts (no prompts)
- `--trust-all-dependencies` or glob-based `--trust-dependencies='*'` CLI flag
- Configuration of the above two behaviors via `bunfig.toml`

Should these be split into additional issues?

**Comment by Jarred-Sumner** [comment]
Fixed by @dylan-conway and @paperdave  in #7132

**Comment by Jarred-Sumner** [comment]
We will ship a follow-up that warns on skipped postinstall scripts in a future release

**Comment by brianmhunt** [comment]
@Jarred-Sumner just a ping re [your comment above](https://github.com/oven-sh/bun/issues/4959#issuecomment-1805670960):

> ... We will add a flag to allow all as well. We are aiming to ship it by Monday.

Not to pester, just making sure it didn't get lost in the shuffle.  Thank you!

**Comment by balloman** [comment]
Asking if we ever got the flag for the post install script? I don't see it in the documentation anywhere. Is there an issue tracking it somewhere?

It seems that --trust would be an option, but that doesn't actually do anything, you still need to manually trust the packages afterwards. (And in fact, using --trust doesn't even give you the warning, so you would have no idea)
